x/stakeibc/keeper/unbonding_records.go

```diff
@@ -188,11 +177,11 @@ func (k Keeper) SweepAllUnbondedTokens(ctx sdk.Context) {
 		for _, unbondingRecord := range unbondingRecords {
 			k.Logger(ctx).Info(fmt.Sprintf("processing unbondingRecord %v", unbondingRecord.Id))
 
-			// total amount of tokens to be swept
-
 			// iterate through all host zone unbondings and process them if they're ready to be swept
-			// TODO() index into the HostZoneUnbonding map with chainID rather than iterating and checking chainID equality
-			unbonding := unbondingRecord.HostZoneUnbondings[zoneInfo.ChainId]
+			unbonding, found := k.RecordsKeeper.GetHostZoneUnbondingByChainId(ctx, unbondingRecord.Id, zoneInfo.ChainId)
+			if !found {
+				return sdkerrors.Wrapf(types.ErrInvalidHostZone, "host zone not found in unbondings: %s", zoneInfo.ChainId)
```

Comment:
If we `return` on `!found` here, a single broken hostZoneUnbonding will prevent processing any unprocessed unbondingRecord across all unprocessed host zones I believe

Should we `continue` instead? 

I lean towards return, but this raises a broader design question -- do we want the system to continue working at large if a single piece breaks, or stop altogether as soon as any error is detected. I think the latter.

Reply:
Hmm this is a good question. We use continue in some places (e.g. a few lines below, if we can't find the host zone) and return errors in others (e.g. if the light client can't return the time), so we don't have a consistent pattern right now.

I think I also slightly lean towards returning errors here, given it makes the code paths much easier to reason about (it values safety over liveness).

Reply:
Given this, I'll update the host zone check to return an error as well

Reply:
What's a little bit odd here is all of the host zones _before_ the zone with the corrupted record state will process ICAs, all of the zones after will not. This feels pretty arbitrary

Reply:
Yeah, hard to make operations atomic when you're doing crosschain logic [](edited-with-Axolo)

Reply:
Agreed the latter is safer + easier to reason about. The only way I can imagine us being able to deal with the issue raised in your last comment is to aggregate all the processing that needs to be done, then process it in a batch